internal/controller/authenticator/jwtcachefiller/jwtcachefiller_test.go

```diff
@@ -144,66 +149,340 @@ func TestController(t *testing.T) {
+			Status:             "False",
+			ObservedGeneration: observedGeneration,
+			LastTransitionTime: time,
+			Reason:             "InvalidDiscoveryProbe",
+			Message:            fmt.Sprintf(`could not perform oidc discovery on provider issuer: Get "%s/.well-known/openid-configuration": dial tcp %s: connect: connection refused`, issuer, parsed.Host),
+		}
+	}
+
+	happyJWKSURLValid := func(time metav1.Time, observedGeneration int64) metav1.Condition {
+		return metav1.Condition{
+			Type:               "JWKSURLValid",
+			Status:             "True",
+			ObservedGeneration: observedGeneration,
+			LastTransitionTime: time,
+			Reason:             "Success",
+			Message:            "jwks_uri is a valid URL",
+		}
+	}
+	unknownJWKSURLValid := func(time metav1.Time, observedGeneration int64) metav1.Condition {
+		return metav1.Condition{
+			Type:               "JWKSURLValid",
+			Status:             "Unknown",
+			ObservedGeneration: observedGeneration,
+			LastTransitionTime: time,
+			Reason:             "UnableToValidate",
+			Message:            "unable to validate; other issues present",
+		}
+	}
+	sadJWKSURLValidParseURI := func(issuer string, time metav1.Time, observedGeneration int64) metav1.Condition {
+		return metav1.Condition{
+			Type:               "JWKSURLValid",
+			Status:             "False",
+			ObservedGeneration: observedGeneration,
+			LastTransitionTime: time,
+			Reason:             "InvalidProviderJWKSURL",
+			Message:            `could not parse provider jwks_uri: parse "` + issuer + `": invalid character " " in host name`,
+		}
+	}
+	sadJWKSURLValidScheme := func(issuer string, time metav1.Time, observedGeneration int64) metav1.Condition {
+		return metav1.Condition{
+			Type:               "JWKSURLValid",
+			Status:             "False",
+			ObservedGeneration: observedGeneration,
+			LastTransitionTime: time,
+			Reason:             "InvalidProviderJWKSURLScheme",
+			Message:            `jwks_uri ` + issuer + ` has invalid scheme, require 'https'`,
+		}
+	}
+
+	allHappyConditionsSuccess := func(issuer string, someTime metav1.Time, observedGeneration int64) []metav1.Condition {
+		return status.SortConditionsByType([]metav1.Condition{
+			happyAuthenticatorValid(someTime, observedGeneration),
+			happyDiscoveryURLValid(someTime, observedGeneration),
+			happyIssuerURLValid(someTime, observedGeneration),
+			happyJWKSURLValid(someTime, observedGeneration),
+			happyReadyCondition(someTime, observedGeneration),
+			happyTLSConfigurationValid(someTime, observedGeneration),
+		})
+	}
+
 	tests := []struct {
-		name                             string
-		cache                            func(*testing.T, *authncache.Cache, bool)
-		syncKey                          controllerlib.Key
-		jwtAuthenticators                []runtime.Object
-		wantClose                        bool
-		wantErr                          testutil.RequireErrorStringFunc
-		wantLogs                         []string
+		name  string
```

Comment:
Do you have any test coverage to show that the observed generation value on the status conditions is correct? And that the merge of new statuses on top of old statuses is happening, with timestamps also being updated appropriately?

Reply:
Added:
```
updateStatus: when updateStatus() is called with different original and updated conditions, it will update the conditions
```

Reply:
But nothing on the timestamp... I've been using "frozenNow" everywhere. 😢

Reply:
Added `Sync(): will update timestamps on relevant statuses`.